catalog/dags/popularity/dag_factory.py

```diff
@@ -0,0 +1,181 @@
+
+Once the constants have been updated, the DAG will trigger a `batched_update`
+DagRun for each provider of this media_type that is configured to support popularity
+data. The batched update recalculates standardized popularity scores for all
+records, using the new constant. When the updates are complete, all records have
+up-to-date popularity data. This DAG can be run concurrently with data refreshes
+and regular ingestion.
+
+
+You can find more background information on this process in the following
+implementation plan:
+
+- [[Implementation Plan] Decoupling Popularity Calculations from the Data Refresh](
+https://docs.openverse.org/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.html)
+"""
+import datetime
+import logging
+import os
+
+from airflow import DAG
+from airflow.decorators import task
+from airflow.operators.trigger_dagrun import TriggerDagRunOperator
+from popularity.popularity_refresh_types import (
+    POPULARITY_REFRESH_CONFIGS,
+    PopularityRefresh,
+)
+from popularity.refresh_popularity_metrics_task_factory import (
+    create_refresh_popularity_metrics_task_group,
+)
+
+from common import slack
+from common.constants import DAG_DEFAULT_ARGS, POSTGRES_CONN_ID
+from common.popularity import sql
+from database.batched_update.constants import DAG_ID as BATCHED_UPDATE_DAG_ID
+
+
+logger = logging.getLogger(__name__)
+
+
+@task
+def notify_slack(text: str, media_type: str, dag_id: str) -> None:
+    slack.send_message(
+        text,
+        username=f"{media_type} Popularity Refresh",
+        icon_emoji=":database:",
+        dag_id=dag_id,
+    )
+
+
+@task
+def get_providers_update_confs(
+    postgres_conn_id: str, popularity_refresh: PopularityRefresh
+):
+    """
+    Build a list of DagRun confs for each provider of this media type. The confs will
+    be used by the `batched_update` DAG to perform a batched update of all existing
+    records, to recalculate their standardized_popularity with the new popularity
+    constant. Providers that do not support popularity data are omitted.
+    """
+    # For the media type, get a list of the providers who support popularity data
+    providers = sql.get_providers_with_popularity_data_for_media_type(
+        postgres_conn_id, popularity_refresh.media_type
+    )
+
+    # Once the new popularity constants are calculated, they will automatically be used
+    # to calculate popularity scores at ingestion. The popularity refresh need only
+    # update the scores for records which were last updated before this time.
+    last_updated_time = datetime.datetime.utcnow()
+
+    # For each provider, create a conf that will be used by the batched_update to
+    # refresh standardized popularity scores.
+    return [
+        {
+            # Uniquely identify the query
+            "query_id": (
+                f"{provider}_popularity_refresh_{last_updated_time.strftime('%Y%m%d')}"
```

Comment:
Do you think this is fine enough resolution for this DAG? There wouldn't be a case where we'd run it twice in one day potentially?

Reply:
The query_id is used for building the temp table in the triggered `batched_update` runs, which is dropped when the update is successful. If it's not successful, we should be managing that by clearing tasks or by manually triggering a new one with the `resume_update` param to use the existing temp table.

This DAG has `max_active_runs=1`, so I think the only way to get a collision would be to start a popularity refresh, fail it and at least one of the triggered `batched_update`s, and then retry the `popularity_refresh` DAG on the same day. In that case a failure seems fine, since that's not the intended way to handle errors for batched updates anyway.